internal/user_validator.go

```diff
@@ -90,24 +95,62 @@ func (i *ValidateUser) Setup(ctx context.Context) error {
 
 func decodePgpKey(pgpKey string) (*openpgp.Entity, error) {
 	pgpKey = strings.TrimRight(pgpKey, " \n\r")
-	if strings.Contains(pgpKey, " ") {
-		return nil, fmt.Errorf("PGP key has spaces in it")
+	pgpKey = strings.TrimSpace(pgpKey)
+
+	var errArmoredKey = errors.New("ASCII-armored PGP keys are not supported")
+
+	keyArmor := func(anchor string) string {
+		return fmt.Sprintf("-----%s %s-----", anchor, openpgp.PublicKeyType)
 	}
+	keyArmorStart := keyArmor("BEGIN")
 
-	suffix := ""
-	for i := 0; i < strings.Count(pgpKey, "="); i++ {
-		suffix += "="
+	if strings.HasPrefix(pgpKey, keyArmorStart[:strings.Index(keyArmorStart, " ")]) {
+		return nil, errArmoredKey
 	}
-	if !strings.HasSuffix(pgpKey, suffix) {
-		return nil, fmt.Errorf("Equals signs are not add the end")
+
+	if strings.Contains(pgpKey, " ") {
+		return nil, fmt.Errorf("given PGP key cannot contain spaces")
 	}
 
 	data, err := base64.StdEncoding.DecodeString(pgpKey)
 	if err != nil {
-		return nil, fmt.Errorf("error decoding given PGP key: %w", err)
+		// Save the original Base64 decoder error,
+		// to return if an error is not related to
+		// ASCII armor parsing or validation.
+		decodeErr := err
+
+		pgpKey = fmt.Sprintf("%s\n\n%s\n%s", keyArmorStart, pgpKey, keyArmor("END"))
+		block, err := armor.Decode(strings.NewReader(pgpKey))
+		if err != nil {
+			return nil, fmt.Errorf("error decoding given ASCII-armored PGP key: %w", err)
+		}
+
+		// Drain the Reader buffer, which causes the CRC24
+		// checksum to be computed for the given ASCII armor.
+		_, err = io.Copy(ioutil.Discard, block.Body)
+		if err != nil {
+			if _, ok := err.(pgperr.StructuralError); ok {
+				return nil, fmt.Errorf("error decoding given ASCII-armored PGP key: %w", err)
+			}
+			return nil, fmt.Errorf("error decoding given PGP key: %w", decodeErr)
```

Comment:
**To the reviewer:**

Internally, the ASCII armor decoder also attempts to decode the Base64-encoded data, however when it fails to do so, then it would return not-so-useful **io.EOF**, and even though still cryptic, the Base64 decoder's errors are still more useful, so we return it instead.

Reply:
ah now i understand why you keep the original decoder error. nice!